src/AbstractActionReader.ts

```diff
@@ -1,38 +1,34 @@
 import * as Logger from "bunyan"
-import { Block, BlockInfo } from "./interfaces"
+import { ActionReaderConfig, Block, BlockInfo, BlockMeta } from "./interfaces"
 
 /**
  * Reads blocks from a blockchain, outputting normalized `Block` objects.
  */
 export abstract class AbstractActionReader {
+  public startAtBlock: number
   public headBlockNumber: number = 0
   public currentBlockNumber: number
-  public isFirstBlock: boolean = true
+  public currentBlockMeta: BlockMeta | null = null
+  protected onlyIrreversible: boolean
+  protected maxHistoryLength: number
   protected currentBlockData: Block | null = null
   protected lastIrreversibleBlockNumber: number = 0
   protected blockHistory: Block[] = []
   protected log: Logger
   private isFirstRun: boolean = true
 
- /**
-  * @param startAtBlock      For positive values, this sets the first block that this will start at. For negative
-  *                          values, this will start at (most recent block + startAtBlock), effectively tailing the
-  *                          chain. Be careful when using this feature, as this will make your starting block dynamic.
-  *
-  * @param onlyIrreversible  When false (default), `getHeadBlockNumber` will load the most recent block number. When
-  *                          true, `getHeadBlockNumber` will return the block number of the most recent irreversible
-  *                          block. Keep in mind that `getHeadBlockNumber` is an abstract method and this functionality
-  *                          is the responsibility of the implementing class.
-  *
-  * @param maxHistoryLength  This determines how many blocks in the past are cached. This is used for determining
-  *                          block validity during both normal operation and when rolling back.
-  */
-  constructor(
-    public startAtBlock: number = 1,
-    protected onlyIrreversible: boolean = false,
-    protected maxHistoryLength: number = 600,
-  ) {
-    this.currentBlockNumber = startAtBlock - 1
+  constructor(config?: ActionReaderConfig) {
+    const configWithDefaults = {
```

Comment:
Because then that leaves us to choose either of two inconveniences:

1. `this.startAtBlock`, `this.onlyIrreversible`, and `this.maxHistoryLength` will change to `this.config.startAtBlock`, `this.config.onlyIrreversible`, and `this.config.maxHistoryLength`, which is inconsistent, annoying to remember, and has no semantic value.

2. Or, the values are copied, and they now exist in two places and can potentially get out of sync, all sorts of problems etc.

It seems to me that using the config object to bootstrap the object properties is a much cleaner approach once you get past the constructor.

Reply:
If we decide to move something into config that existed as a property before, this would also necessitate a breaking change. If we  leave them as properties and re-assign in the constructor, then it's not necessarily a breaking change.

Reply:
How would this be a breaking change, do other libraries depend on the ability to read the values from this object?

Maybe move to get? so we can still expose the values but also avoid unintentional variable writing?  That would make it non-breaking but also avoid copying, avoiding out of sync and un-intentional overwrites.

```
public get startAtBlock(): string {
    return this.config.startAtBlock;
}
etc...

// usage everywhere else
this.startAtBlock

actionReaderInstance.startAtBlock //now without being able to be overwritten
```

Reply:
But now it can't be re-assigned (intentionally) without going to config. Why introduce this complexity to avoid a few extra lines in the constructor? If the extra lines are a problem, we could do something like:

```js
for (const key of Object.keys(config)) {
  this[key] = config[key]
}
```
But I'd prefer to stay away from all this magic and just be explicit.

Reply:
> do other libraries depend on the ability to read the values from this object?
Yes, namely the implementing libraries (demux-js-eos, demux-js-postgres, etc)

Reply:
The goal isn't to have a single source of truth for the config- the class properties themselves are already that single source of truth. The reason we're moving to config objects is because javascript lacks named arguments. Because of this, developers were forced to fill every positional argument, even if they need default values for everything but the last argument. Lifting this burden also allows us to add more "named arguments" freely, without having to consider further detriment to the developer experience.

`startAtBlock`,  `onlyIrreversible`, and `maxHistoryLength` were mutable before this PR, and changing the mutability of these values is outside of the intention of this PR. `startAtBlock` in particular needs to be able to change, due to how "tailing" the chain works via negative values. I think part of the problem here is because the name "config" has connotations that aren't intended. I still think it's the best name, but I'm open to others.

Reply:
While javascript lacks protection, Typescript doesn't. `startAtBlock`, `onlyIrreversible`, and `maxHistoryLength` are respectively public, protected, and protected properties, before this PR, and in this PR- that hasn't changed.

Reply:
> I mean your configWithDefaults is a const then we assign those const values to a non-const based variable meaning anything in the system could overwrite those values leading to un-intentional side effects.

It seems to me you're conflating variable scope immutability with object immutability, which are different.

Reply:
If you want a name that may remove some of the `config` confusion, `params` might work

Reply:
The term `options` came to me over the holidays. I think that might be better than config (and params).